packages/@aws-cdk/aws-ec2/lib/volume.ts

```diff
@@ -210,4 +213,351 @@ export enum EbsDeviceVolumeType {
+    return result;
+  }
+
+  public grantDetachVolume(grantee: IGrantable, instances?: IInstance[]): Grant {
+    const result = Grant.addToPrincipal({
+      grantee,
+      actions: [ 'ec2:DetachVolume' ],
+      resourceArns : this.collectGrantResourceArns(instances),
+    });
+    // Note: Access to the encryption key is not required. The service having access to it is sufficient.
+    return result;
+  }
+
+  private collectGrantResourceArns(instances?: IInstance[]): string[] {
+    const stack = Stack.of(this);
+    const resourceArns: string[] = [
+      `arn:aws:ec2:${stack.region}:${stack.account}:volume/${this.volumeId}`,
+    ];
+    if (instances) {
+      instances.forEach(instance => resourceArns.push(`arn:aws:ec2:${stack.region}:${stack.account}:instance/${instance?.instanceId}`));
+    } else {
+      resourceArns.push(`arn:aws:ec2:${stack.region}:${stack.account}:instance/*`);
+    }
+    return resourceArns;
+  }
+}
+
+/**
+ * Creates a new EBS Volume in AWS EC2.
+ */
+export class Volume extends VolumeBase {
+  /**
+   * Import an existing EBS Volume into the Stack.
+   *
+   * @param scope the scope of the import.
+   * @param id    the ID of the imported Volume in the construct tree.
+   * @param attrs the attributes of the imported Volume
+   */
+  public static fromVolumeAttributes(scope: Construct, id: string, attrs: VolumeAttributes): IVolume {
+    class Import extends VolumeBase {
+      public readonly volumeId = attrs.volumeId;
+      public readonly availabilityZone = attrs.availabilityZone;
+      public readonly encryptionKey = attrs.encryptionKey;
+    }
+    // Check that the provided volumeId looks like it could be valid.
+    if (!Token.isUnresolved(attrs.volumeId) && !/^vol-[0-9a-fA-F]+$/.test(attrs.volumeId)) {
+      throw new Error('`volumeId` does not match expected pattern. Expected `vol-<hexadecmial value>` (ex: `vol-05abe246af`) or a Token');
+    }
+    return new Import(scope, id);
+  }
+
+  public readonly volumeId: string;
+  public readonly availabilityZone: string;
+  public readonly encryptionKey?: IKey;
+
+  constructor(scope: Construct, id: string, props: VolumeProps) {
+    super(scope, id, {
+      physicalName: props.volumeName,
+    });
+
+    this.validateProps(props);
+
+    const resource = new CfnVolume(this, 'Resource', {
+      availabilityZone: props.availabilityZone,
+      autoEnableIo: props.autoEnableIo,
+      encrypted: props.encrypted,
+      kmsKeyId: props.encryptionKey?.keyArn,
+      iops: props.iops,
+      multiAttachEnabled: props.enableMultiAttach ?? false,
+      size: props.size?.toGibibytes({rounding: SizeRoundingBehavior.FAIL}),
+      snapshotId: props.snapshotId,
+      volumeType: props.volumeType ?? EbsDeviceVolumeType.GENERAL_PURPOSE_SSD,
+    });
+
+    this.volumeId = resource.ref;
+    this.availabilityZone = props.availabilityZone;
+    this.encryptionKey = props.encryptionKey;
+
+    if (this.encryptionKey) {
```

Comment:
The role being added here might not be 100% correct. I created it by mimicing the service KMS key for EBS, but it's weird to me that you don't need any special IAM permissions to use an EBS device that has a customer managed KMS key with this policy.
AFAIK, the whole point of using a customer-managed key instead of the service key is that there's an extra layer of IAM permissions required to use the service.

I'm going to seek out an expert. Will update with the result.

Reply:
This has been sorted out by going through a tonne of documentation, and a lot of trial and error.